### PR TITLE
Fix TOB-4 nonces unique per token

### PR DESCRIPTION
**Pull request by wjuan-mob**
- Add filter in combine mint config and combine mint tx for duplicate nonces across different tokens
- Add tests for filtering duplicate nonces across different tokens
- Migrate LedgerDB to store nonces per token

### Motivation

This is to address (https://github.com/mobilecoinfoundation/mobilecoin/issues/2323)
This specifically addresses: https://github.com/mobilecoinfoundation/mobilecoin/issues/2325
and https://github.com/mobilecoinfoundation/mobilecoin/issues/2322

### Future Work
Not addressing enclave issues in this PR
https://github.com/mobilecoinfoundation/mobilecoin/issues/2324

[Soundtrack of this PR](https://www.youtube.com/watch?v=DFu8P7teOt8)

### CODE REVIEW

File: fog/sample-paykit/src/cached_tx_data/mod.rs [hunk 1]
```diff
@@ -66,7 +65,15 @@ const TELEMETRY_NUM_TXOS_KEY: Key = telemetry_static_key!("num-txos");
...
-const SUBADDRESS_LOW_RANGE: RangeInclusive<u64> = 0..=DEFAULT_SUBADDRESS_INDEX;
+///
+/// Note: We also put the historical change subaddress of 1 in this range.
+/// This is because we ran the test client as a canary for several months using
+/// 1 as the change subaddress, so those accounts now have many subaddress index
+/// 1 TxOuts. Adding this to the searched range means that they don't lose all
+/// this money, and also silences the warnings when they get TxOuts from fog on
+/// "unexpected" subaddresses.
+const HISTORICAL_CHANGE_SUBADDRESS_INDEX: u64 = 1;
+const SUBADDRESS_LOW_RANGE: RangeInclusive<u64> = 0..=HISTORICAL_CHANGE_SUBADDRESS_INDEX;
```
Comment (garbageslam) [inline fog/sample-paykit/src/cached_tx_data/mod.rs:76]: Why is this showing up in the diff?

File: ledger/db/src/mint_config_store.rs [hunk 2]
```diff
@@ -28,8 +28,8 @@ use mc_util_serial::{decode, encode, Message};
 // LMDB Database names.
 pub const ACTIVE_MINT_CONFIGS_BY_TOKEN_ID_DB_NAME: &str =
     "mint_config_store:active_mint_configs_by_token_id";
-pub const BLOCK_INDEX_BY_MINT_CONFIG_TX_NONCE_DB_NAME: &str =
-    "mint_config_store:block_index_by_mint_config_tx_nonce";
+pub const BLOCK_INDEX_BY_MINT_CONFIG_TX_NONCE_AND_TOKEN_ID_DB_NAME: &str =
+    "mint_config_store:block_index_by_mint_config_tx_nonce_and_token_id";
```
Comment (garbageslam) [inline ledger/db/src/mint_config_store.rs:32]: I think this isn't going to work, we need to actually create a migration don't we? We're going to turn on the nonces in prod soon, we can't just delete the old tables?

File: mint-auditor/migrations/2022-05-24-214504_create_initial/up.sql [hunk 3]
```diff
@@ -35,7 +35,7 @@ CREATE TABLE mint_config_txs (
     -- The token id this mint config tx is for.
     token_id BIGINT NOT NULL,
     -- The nonce, as hex-encoded bytes.
-    nonce_hex VARCHAR(128) NOT NULL UNIQUE,
+    nonce_hex VARCHAR(128) NOT NULL,
```
Comment (garbageslam) [inline mint-auditor/migrations/2022-05-24-214504_create_initial/up.sql:38]: These mint-auditor changes need to be coordinated with @david1alvarez 

This is not the real mint-auditor anymore, this is just here for testing.
See README of mint-auditor crate

Comment (garbageslam) [reply mint-auditor/migrations/2022-05-24-214504_create_initial/up.sql:38]: I'm not sure if it's necessary to change the mint auditor at all in this PR, will it actually break if we don't change it?

File: fog/sample-paykit/src/cached_tx_data/mod.rs (hunk 1 above)
Comment (wjuan-mob) [reply fog/sample-paykit/src/cached_tx_data/mod.rs:76]: I cherry picked the PR on release/v2 into this one to master. I think this got pulled in from there. I'm not sure if this is still correct or not. The comment seems to make sense?

File: mint-auditor/migrations/2022-05-24-214504_create_initial/up.sql (hunk 3 above)
Comment (wjuan-mob) [reply mint-auditor/migrations/2022-05-24-214504_create_initial/up.sql:38]: test_find_unaudited_mint_tx_by_nonce seems to fail when I remove this change. I'll coordinate with @david1alvarez.

File: fog/sample-paykit/src/cached_tx_data/mod.rs (hunk 1 above)
Comment (eranrund) [reply fog/sample-paykit/src/cached_tx_data/mod.rs:76]: It might be correct but it is unrelated to this PR :)

File: mint-auditor/migrations/2022-05-24-214504_create_initial/up.sql (hunk 3 above)
Comment (eranrund) [reply mint-auditor/migrations/2022-05-24-214504_create_initial/up.sql:38]: It will break if at some point the nonce is not unique, the unique constraint now needs to be on `(token_id, nonce_hex)`. If we are only planning on using the mint-auditor here for internal testing then that's not a problem, but the reserve auditor should alter the unique constraint.

File: fog/sample-paykit/src/cached_tx_data/mod.rs (hunk 1 above)
Comment (garbageslam) [reply fog/sample-paykit/src/cached_tx_data/mod.rs:76]: Is there a reason to target master instead of release/v2?

I think we would release this in v2.1 if we can

Comment (garbageslam) [reply fog/sample-paykit/src/cached_tx_data/mod.rs:76]: maybe its better to wait

Comment (eranrund) [reply fog/sample-paykit/src/cached_tx_data/mod.rs:76]: Seems to me that targeting `master` makes sense since we definitely need it in master, and only maybe we also want it in release/v2

File: ledger/db/src/mint_config_store.rs (hunk 2 above)
Comment (wjuan-mob) [reply ledger/db/src/mint_config_store.rs:32]: I attempted the migration, and tested it locally. I think it should be ok to delete the old tables and rebuild in this migration path.

Comment (garbageslam) [reply ledger/db/src/mint_config_store.rs:32]: okay sounds good to me :)

File: mint-auditor/src/db/models/audited_mint.rs [hunk 4]
```diff
@@ -387,12 +424,13 @@ mod tests {
...
 
-        // No mismatching pairs were found.
+        // 3 attempts to match returned not found errors, which register as as a
+        // mismatch.
         assert_eq!(
             Counters::get(&conn)
                 .unwrap()
                 .num_mismatching_mints_and_deposits(),
-            0
+            3
```
Comment (eranrund) [inline mint-auditor/src/db/models/audited_mint.rs:433]: I am not sure this is right. What made you change it?

File: mint-auditor/src/db/models/audited_mint.rs [hunk 5]
```diff
@@ -503,12 +541,9 @@ mod tests {
         insert_mint_tx_from_deposit(&deposit, &conn, &mut rng);
 
         config.tokens[0].token_id = TokenId::from(123);
-
-        assert!(matches!(
-            AuditedMint::try_match_deposit_with_mint(&deposit, &config, &conn),
-            Err(Error::DepositAndMintMismatch(_))
-        ));
-
+        let result = AuditedMint::try_match_deposit_with_mint(&deposit, &config, &conn);
```
Comment (eranrund) [inline mint-auditor/src/db/models/audited_mint.rs:544]: Can you please explain why a bunch of tests got changed? I don't see why they should be producing different results all of the sudden.

File: mint-auditor/src/db/models/gnosis_safe_deposit.rs [hunk 6]
```diff
@@ -141,11 +141,13 @@ impl GnosisSafeDeposit {
 
     /// Attempt to find a [GnosisSafeDeposit] that has a given nonce and no
     /// matching entry in the `audited_mints` table.
-    pub fn find_unaudited_deposit_by_nonce(
+    pub fn find_unaudited_deposit_by_nonce_and_token_address(
```
Comment (eranrund) [inline mint-auditor/src/db/models/gnosis_safe_deposit.rs:144]: As mentioned on a previous comment, I don't think we can encounter the same nonce twice on the Gnosis side, since it is derived from the ethereum tx hash - which is guaranteed to be unique.

Does that make sense? Please double check this assumption :)

File: mint-auditor/src/db/models/mint_config_tx.rs [hunk 7]
```diff
@@ -350,15 +350,18 @@ mod tests {
...
         let token_id1 = TokenId::from(1);
+        let token_id2 = TokenId::from(2);
 
         let conn = mint_auditor_db.get_conn().unwrap();
-
+        let mut rng = mc_util_test_helper::get_seeded_rng();
+        let mut rng2 = mc_util_test_helper::get_seeded_rng();
         let (mint_config_tx1, _signers) = create_mint_config_tx_and_signers(token_id1, &mut rng);
         let (mint_config_tx2, _signers) = create_mint_config_tx_and_signers(token_id1, &mut rng);
+        let (mint_config_tx1_tkn2, _signers) =
```
Comment (eranrund) [inline mint-auditor/src/db/models/mint_config_tx.rs:363]: Instead of using two rngs, would it be possible to clone the original mint_config_tx_1 and change the token_id in it? I think that is a bit easier to follow

File: mint-auditor/src/db/models/mint_tx.rs [hunk 8]
```diff
@@ -200,6 +202,7 @@ impl MintTx {
 
 #[cfg(test)]
 mod tests {
+
```
Comment (eranrund) [inline mint-auditor/src/db/models/mint_tx.rs:205]:
```suggestion
```

File: mint-auditor/src/db/models/mint_tx.rs [hunk 9]
```diff
@@ -230,72 +233,184 @@ mod tests {
...
+        let mut mint_tx2_tkn2 = create_mint_tx(token_id2, &signers2, 100, &mut rng2);
 
         mint_tx1.prefix.nonce = hex::decode(&deposit1.expected_mc_mint_tx_nonce_hex()).unwrap();
         mint_tx2.prefix.nonce = hex::decode(&deposit2.expected_mc_mint_tx_nonce_hex()).unwrap();
+        mint_tx1_tkn2.prefix.nonce =
+            hex::decode(&deposit3.expected_mc_mint_tx_nonce_hex()).unwrap();
+        mint_tx2_tkn2.prefix.nonce =
+            hex::decode(&deposit4.expected_mc_mint_tx_nonce_hex()).unwrap();
+
+        //The nonces should be the same.
```
Comment (eranrund) [inline mint-auditor/src/db/models/mint_tx.rs:268]:
```suggestion
        // The nonces should be the same.
```

File: mint-auditor/src/db/models/mint_tx.rs [hunk 10]
```diff
@@ -230,72 +233,184 @@ mod tests {
...
+                &hex::encode(&mint_tx1.prefix.nonce),
+                &conn
+            )
+            .unwrap()
+            .unwrap(),
             sql_mint_tx1
         );
-
-        assert!(MintTx::find_unaudited_mint_tx_by_nonce(
+        //This has not been inserted, and should not be found.
```
Comment (eranrund) [inline mint-auditor/src/db/models/mint_tx.rs:318]:
```suggestion
        // This has not been inserted, and should not be found.
```

File: mint-auditor/src/db/models/mint_tx.rs [hunk 11]
```diff
@@ -230,72 +233,184 @@ mod tests {
...
+        assert!(MintTx::find_unaudited_mint_tx_by_nonce_and_token_id(
+            token_id1,
             &hex::encode(&mint_tx2.prefix.nonce),
             &conn
         )
         .unwrap()
         .is_none());
+        //This is for a different token_id, so should not be found even though it
```
Comment (eranrund) [inline mint-auditor/src/db/models/mint_tx.rs:326]:
```suggestion
        // This is for a different token_id, so should not be found even though it
```

File: mint-auditor/src/db/models/mint_tx.rs [hunk 12]
```diff
@@ -230,72 +233,184 @@ mod tests {
...
+        // shares a nonce.
+        assert!(MintTx::find_unaudited_mint_tx_by_nonce_and_token_id(
+            token_id2,
+            &hex::encode(&mint_tx1_tkn2.prefix.nonce),
+            &conn
+        )
+        .unwrap()
+        .is_none());
+        //This is for a different token_id, so should not be found even though it
```
Comment (eranrund) [inline mint-auditor/src/db/models/mint_tx.rs:335]:
```suggestion
        // This is for a different token_id, so should not be found even though it
```

File: mint-auditor/src/db/models/mint_tx.rs [hunk 13]
```diff
@@ -230,72 +233,184 @@ mod tests {
...
+            MintTx::find_unaudited_mint_tx_by_nonce_and_token_id(
+                token_id1,
+                &hex::encode(&mint_tx2.prefix.nonce),
+                &conn
+            )
+            .unwrap()
+            .unwrap(),
             sql_mint_tx2
         );
+        //This is for a different token_id, so should not be found even though it
```
Comment (eranrund) [inline mint-auditor/src/db/models/mint_tx.rs:369]:
```suggestion
        // This is for a different token_id, so should not be found even though it
```

File: mint-auditor/src/db/models/mint_tx.rs [hunk 14]
```diff
@@ -230,72 +233,184 @@ mod tests {
...
+        // shares a nonce.
+        assert!(MintTx::find_unaudited_mint_tx_by_nonce_and_token_id(
+            token_id2,
+            &hex::encode(&mint_tx1_tkn2.prefix.nonce),
+            &conn
+        )
+        .unwrap()
+        .is_none());
+        //This is for a different token_id, so should not be found even though it
```
Comment (eranrund) [inline mint-auditor/src/db/models/mint_tx.rs:378]:
```suggestion
        // This is for a different token_id, so should not be found even though it
```

**Review by eranrund — CHANGES_REQUESTED** [review]
Nice work. I think the most interesting thing I found is the stuff around the expected nonce being unique on the gnosis side, since it is derived from the ethereum tx hash (which has to be unique if ethereum is not broken). If I haven't missed something by that assumption then it should make some of the changes unnecessary and simplifies things a bit.
My other main concern is around the NotFound case when counting errors, which I believe is wrong and is likely the cause of a lot of test assertions changing unnecessarily.
Thank you!

File: mint-auditor/src/db/models/audited_mint.rs (hunk 5 above)
Comment (wjuan-mob) [reply mint-auditor/src/db/models/audited_mint.rs:544]: They're producing different results because they're throwing different errors after the changes I made. Instead of getting caught inside verify, noting that the tokenid is wrong when you look up this mint and does not correspond to the deposit, you fail to find the mint at all.

File: mint-auditor/src/db/models/audited_mint.rs [hunk 15]
```diff
@@ -147,12 +159,29 @@ impl AuditedMint {
...
+                    )));
+                }
+                let eth_token = eth_tokens[0];
 
                 // See if we can find a GnosisSafeDeposit that matches the nonce and has not
                 // been associated with a mint.
-                let deposit =
-                    GnosisSafeDeposit::find_unaudited_deposit_by_nonce(mint_tx.nonce_hex(), conn)?
-                        .ok_or(Error::NotFound)?;
+                let deposit = GnosisSafeDeposit::find_unaudited_deposit_by_nonce_and_token_address(
```
[the remark this answers is not in the document]
Comment (wjuan-mob) [reply mint-auditor/src/db/models/audited_mint.rs:179]: Yes. I do see that this makes sense. I think I hadn't realized that the ethereum side nonce had to be unique. I can back out these changes on the gnosis safe side given that, and that should also resolve the question of: why did all these tests change.

File: mint-auditor/src/db/models/audited_mint.rs [hunk 16]
```diff
@@ -102,7 +112,9 @@ impl AuditedMint {
             Err(Error::DepositAndMintMismatch(_)) => {
                 Counters::inc_num_mismatching_mints_and_deposits(conn)?;
             }
-
+            Err(Error::NotFound) => {
```
[the remark this answers is not in the document]
Comment (wjuan-mob) [reply mint-auditor/src/db/models/audited_mint.rs:115]: The thought here was that when there is a deposit, and the mint is on a different token_id, this should result in a NotFound error, because it does not seem to be possible to distinguish by an error where the mint was onto the wrong tokenid, and when there is a coincidental nonce collision with a mint on the 2nd tokenId. After discussion offline, I agree that this should not be a counted error and should instead by monitored by the resulting pair of unaudited mint/deposit. Fixing.

File: mint-auditor/src/db/models/audited_mint.rs [hunk 17]
```diff
@@ -147,12 +159,29 @@ impl AuditedMint {
                         nonce_hex, eth_tx_hash,
                     )));
                 }
+                let eth_tokens = config.get_tokens_by_token_id(&mint_tx.token_id());
```
[the remark this answers is not in the document]
Comment (wjuan-mob) [reply mint-auditor/src/db/models/audited_mint.rs]: I will do this in a separate PR since this doesn't seem specific to this change.

**Review by garbageslam — APPROVED** [review]
thank you! I am okay to see this go in.

I think Eran's comment about cloning the mint tx instead of cloning the Rng might be a good idea, it will be a little easier to follow and less fragile. I think a code comment might also be okay. We need to figure out what the long term plan for this version of mint auditor is.

File: mint-auditor/src/gnosis/config.rs [hunk 18]
```diff
@@ -88,6 +93,13 @@ impl GnosisSafeConfig {
             .find(|safe| safe.safe_addr == *safe_addr)
             .cloned()
     }
+    /// Get an [AuditedToken] by its [TokenId].
+    pub fn get_tokens_by_token_id(&self, token_id: &TokenId) -> Vec<&AuditedToken> {
```
Comment (eranrund) [inline mint-auditor/src/gnosis/config.rs:97]: I think this is no longer used and can be removed.

**Review by eranrund — APPROVED** [review]

File: mint-auditor/src/db/models/mint_config_tx.rs (hunk 7 above)
Comment (wjuan-mob) [reply mint-auditor/src/db/models/mint_config_tx.rs:363]: Done.

File: mint-auditor/src/gnosis/config.rs (hunk 18 above)
Comment (wjuan-mob) [reply mint-auditor/src/gnosis/config.rs]: Agreed.
